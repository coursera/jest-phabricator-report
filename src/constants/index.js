const path = require('path');

const JEST_PHABRICATOR_REPORT_OUTPUT = 'unit-test-phabricator-report.json';

module.exports = {
  ENVIRONMENT_CONFIG_MAP: {
    // Would display the ancestors of a given tests
    // ANCESTOR'NAME > ANCESTOR'S NAME >  TEST
    separator: ' ',
    post_processor: '',
    // report's  output
    JEST_PHABRICATOR_REPORT_OUTPUT,
    // required for producing a valid report.
    buildTargetPHID: '',
    // in case we want to trim the path name (on the report).
    trimBasePath: false,
  },
  DEFAULT_OPTIONS: {
    output: path.join(process.cwd(), JEST_PHABRICATOR_REPORT_OUTPUT),
  },
};
