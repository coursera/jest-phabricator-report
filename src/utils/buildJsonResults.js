// @flow

type Options = {
  buildTargetPHID: string,
  SEPARATOR: string,
  trimBasePath: string,
  basePath: string,
};
type TestsResults = {
  status: 'passed' | 'failed',
  failureMessages: string[],
  title: string,
  ancestorTitles: string[],
  duration: number,
};

type JestTestSuitesResults = {
  numFailingTests: number,
  numPassingTests: number,
  testFilePath: string,
  numPendingTests: number,
  perfStats: { end: number, start: number },
  testResults: TestsResults[],
};
type JestReport = {
  numFailedTestSuites: number,
  numPassedTests: number,
  numTotalTests: number,
  testResults: JestTestSuitesResults[],
};
type PhabricatorUnitTestResult = {
  name: string,
  path: string,
  result: 'pass' | 'fail',
  namespace: string,
  duration: number,
  details: string,
};

type PhabricatorUnitTestsJestReport = {
  buildTargetPHID: string,
  type: 'pass' | 'fail' | 'work',
  unit: PhabricatorUnitTestResult[],
};

type ReportResult = 'pass' | 'fail';

const PHABRICATOR_JEST_UNIT_RESULTS = {
  passed: 'pass',
  failed: 'fail',
};

function getReportResult({ testResults }): ReportResult {
  const { passed, failed } = PHABRICATOR_JEST_UNIT_RESULTS;

  return testResults.some(({ numFailingTests }) => numFailingTests) ? failed : passed;
}

function getResultsPerSuite(path, testResults: TestsResults[], SEPARATOR): PhabricatorUnitTestResult[] {
  return testResults.map(({ status, duration, failureMessages, title: name, ancestorTitles }) => ({
    path,
    name,
    duration: duration / 1000,
    result: PHABRICATOR_JEST_UNIT_RESULTS[status],
    namespace: ancestorTitles.join(SEPARATOR),
    details: failureMessages.join(''),
  }));
}

function trimPath(path, basePath) {
  return path.replace(basePath, '');
}
function getTestResultForPhabricator({ testResults: testSuitResults }, { SEPARATOR, basePath, trimBasePath }) {
  return testSuitResults.reduce(
    (
      memo,
      { testFilePath: name, testResults: results, numFailingTests, perfStats: { end, start } }: JestTestSuitesResults
    ): PhabricatorUnitTestResult[] => {
      const duration = (end - start) / 1000;
      const result = numFailingTests ? 'fail' : 'pass';
      const path = trimBasePath ? trimPath(name, basePath) : name;
      let summary = [];
      if (result === 'fail') {
        summary = [...summary, ...getResultsPerSuite(path, results, SEPARATOR)];
      } else {
        summary.push({
          // we don't use `path` attribute given that the display on Phabricator is not optimal.
          path,
          result,
          duration,
        });
      }
      return [...memo, ...summary];
    },
    []
  );
}

export default function(report: JestReport, options: Options): PhabricatorUnitTestsJestReport {
  const { buildTargetPHID, SEPARATOR, trimBasePath, basePath } = options;

  if (!buildTargetPHID) {
    throw new Error('buildTargetPHID not present');
  }
  return {
    buildTargetPHID,
    type: getReportResult(report),
    unit: getTestResultForPhabricator(report, { SEPARATOR, trimBasePath, basePath }),
  };
}
