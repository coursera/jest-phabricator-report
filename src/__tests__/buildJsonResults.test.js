import buildJsonResults from '../utils/buildJsonResults';
import constants from '../constants/index';

describe('buildJsonResults', () => {
  it('should contain global result of the TestsSuits', () => {
    const noFailingTestsReport = require('../__mocks__/no-failing-tests.json');
    const jsonResults = buildJsonResults(noFailingTestsReport, {
      ...constants.DEFAULT_OPTIONS,
      buildTargetPHID: 'XXXX',
    });
    expect(jsonResults.type).toMatchSnapshot();
  });

  it('should contain the correct shape for unit key', () => {
    const noFailingTestsReport = require('../__mocks__/no-failing-tests.json');
    const jsonResults = buildJsonResults(noFailingTestsReport, {
      ...constants.DEFAULT_OPTIONS,
      buildTargetPHID: 'XXXX',
    });

    expect(jsonResults.unit).toMatchSnapshot();
  });

  it('should fail if buildTargetPHID is not defined', () => {
    const noFailingTestsReport = require('../__mocks__/no-failing-tests.json');

    expect(() => buildJsonResults(noFailingTestsReport, constants.DEFAULT_OPTIONS)).toThrowErrorMatchingSnapshot();
  });

  it('should expand when errors are found', () => {
    const failingTestsReport = require('../__mocks__/failing-tests.json');
    const jsonResults = buildJsonResults(failingTestsReport, {
      ...constants.DEFAULT_OPTIONS,
      buildTargetPHID: 'XXXX',
    });

    expect(jsonResults).toMatchSnapshot();
  });
});
