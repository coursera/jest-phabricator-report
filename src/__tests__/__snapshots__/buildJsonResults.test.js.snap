// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildJsonResults should contain global result of the TestsSuits 1`] = `"pass"`;

exports[`buildJsonResults should contain the correct shape for unit key 1`] = `
Array [
  Object {
    "duration": 0.12,
    "path": "/path/to/test/__tests__/foo.test.js",
    "result": "pass",
  },
]
`;

exports[`buildJsonResults should expand when errors are found 1`] = `
Object {
  "buildTargetPHID": "XXXX",
  "type": "fail",
  "unit": Array [
    Object {
      "details": "[1m[31m  [1m‚óè [1mSample Failing Test ‚Ä∫ Inner ‚Ä∫ Inner Inner ‚Ä∫ Should fail[39m[22m

    foobar
[2m      
      [2mat _callee$ ([2m[0m[36mpath/to/failing.test.js[39m[0m[2m:26:15)[2m
      [2mat tryCatch ([2m[0m[36mnode_modules/regenerator-runtime/runtime.js[39m[0m[2m:64:40)[2m
      [2mat GeneratorFunctionPrototype.invoke [as _invoke] ([2m[0m[36mnode_modules/regenerator-runtime/runtime.js[39m[0m[2m:299:22)[2m
      [2mat GeneratorFunctionPrototype.prototype.(anonymous function) [as next] ([2m[0m[36mnode_modules/regenerator-runtime/runtime.js[39m[0m[2m:116:21)[2m
      [2mat step ([2m[0m[36mpath/to/failing.test.js[39m[0m[2m:2:394)[2m
      [2mat [2m[0m[36mpath/to/failing.test.js[39m[0m[2m:2:554[2m[22m
",
      "duration": 3.93,
      "name": "Should fail",
      "namespace": "Sample Failing Test,Inner,Inner Inner",
      "path": "/path/to/failing.test.js",
      "result": "fail",
    },
  ],
}
`;

exports[`buildJsonResults should fail if buildTargetPHID is not defined 1`] = `"buildTargetPHID not present"`;
